text-motions.c

```diff
@@ -586,25 +586,37 @@ size_t text_function_end_prev(Text *txt, size_t pos) {
 	return text_function_end_direction(txt, pos, -1);
 }
 
-size_t text_block_start(Text *txt, size_t pos) {
+static size_t text_block_direction(Text *txt, size_t pos, int direction) {
 	int nest = 0;
-	char c;
+	char c, target, nest_target;
 	bool instring = false;
 	Iterator it = text_iterator_get(txt, pos);
-	while (text_iterator_byte_prev(&it, &c)) {
+
+	target = (direction > 0 ? '}' : '{');
+	nest_target = (direction > 0 ? '{' : '}');
```

Comment:
While I agree that the current approach has the drawbacks you mention, it works for the common case, is reasonably simple and already exists. It is currently used to:

* highlight matching symbols at the cursor location
* implement the `%` motion when on a special symbol
* implement text objects for blocks enclosed by `{`, `(`, `[` etc.

It will also work in builds without Lua support (whether we want to support these is another question).

This pull requests simply hooks up a few more motions to already existing code. Adding more "semantic" versions of these motions and text objects is a worthwhile, but separate and not completely trivial, goal which should not block this.

Also I might be slightly biased, because these days I primarily work in C like languages.